vllm/model_executor/models/telechat2.py

```diff
@@ -0,0 +1,285 @@
+            config.num_hidden_layers,
+            lambda prefix: TeleChat2DecoderLayer(config=config,
+                                             cache_config=cache_config,
+                                             quant_config=quant_config,
+                                             prefix=f"{prefix}.layers"),
+            prefix=f"{prefix}.layers",
+        )
+        
+    def load_weights(self, weights: Iterable[Tuple[str,
+                                                   torch.Tensor]]) -> Set[str]:
+        stacked_params_mapping = [
+            ('gate_up_proj', 'gate_proj', 0),
+            ('gate_up_proj', 'up_proj', 1),
+        ]
+        params_dict = dict(self.named_parameters())
+        loaded_params: Set[str] = set()
+        total_num_heads = self.config.n_head
+        head_dim = self.config.hidden_size // total_num_heads
+        for name, loaded_weight in weights:
+            #name = name.replace(".h.", ".layers.")
+            if "self_attn.key_value" in name:
+                k_weight = []
+                v_weight = []
+                #name = name.replace(".self_attention.", ".self_attn.")
+                for i in range(total_num_heads):
+                    start =i * head_dim * 2
+                    k_weight.append(loaded_weight[start:start+head_dim,:])
+                    v_weight.append(loaded_weight[start+head_dim:start+2*head_dim:])
+                k_weight = torch.cat(k_weight,dim=0)
+                v_weight = torch.cat(v_weight,dim=0)
+                name = name.replace("key_value", "qkv_proj")
+                param = params_dict[name]
+                weight_loader = param.weight_loader
+                weight_loader(param, k_weight, "k")
+                weight_loader(param, v_weight, "v")
+                loaded_params.add(name)
+            elif "query" in name:
+                name = name.replace("query", "qkv_proj")
+                param = params_dict[name]
+                weight_loader = param.weight_loader
+                weight_loader(param, loaded_weight, "q")
+                loaded_params.add(name)
+            else:
+                for param_name, weight_name, shard_id in stacked_params_mapping:
+                    if weight_name not in name:
+                        continue
+                    name = name.replace(weight_name, param_name)
+                    param = params_dict[name]
+                    weight_loader = param.weight_loader
+                    weight_loader(param, loaded_weight, shard_id)
+                    break
+                else:
+                    param = params_dict[name]
+                    weight_loader = getattr(param, "weight_loader",
+                                        default_weight_loader)
+                    weight_loader(param, loaded_weight)
+        return loaded_params
+    
+
+class TeleChat2ForCausalLM(nn.Module):
+
+    def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
+        super().__init__()
```

Comment:
Can we also inherit from `LlamaForCausalLM`? So that we can only override config, `self.model` and `load_weights`.

Reply:
Thank you for your suggestion!

I just tried inheriting from LlamaForCausalLM, but it seems to cause a memory issue where the VRAM usage doubles when loading the model. I suspect this might be related to how the weights or buffers are handled during initialization. Let me know if you have any insights on this, or if there's a workaround we can try!

Thanks again for your input!

Reply:
Hmmm, this might because we directly initialize the model on GPU in model loader. Let's keep it as it looks now!

Reply:
```suggestion
class TeleChat2ForCausalLM(LlamaForCausalLM):

    def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
        super(LlamaForCausalLM, self).__init__()
```

After consideration, I think we can still have a try to inherit from `LlamaForCausalLM` with keeping `__init__` only inherited from `nn.Module`. So that we can avoid implementing `compute_logits` etc duplicatedly.

Reply:
Thank you for your suggestion! I have modified the code accordingly.